File: src/main/java/org/podval/zenfolio/Bytes.java
package org.podval.zenfolio;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

import javax.activation.DataHandler;
import javax.activation.DataSource;

import java.io.InputStream;
import java.io.OutputStream;
import java.io.ByteArrayInputStream;
import java.io.IOException;


public final class Bytes {

    public static byte[] hash(final byte[] what) throws NoSuchAlgorithmException {
        final MessageDigest digest = MessageDigest.getInstance("SHA-256");
        digest.reset();
        return digest.digest(what);
    }


    public static byte[] concatenate(final byte[] first, final byte[] second) {
        final byte[] result = new byte[first.length + second.length];

        for (int i = 0; i<first.length; i++) {
            result[i] = first[i];
        }

        for (int i = 0; i<second.length; i++) {
            result[first.length+i] = second[i];
        }

        return result;
    }


    public static byte[] readBytes(final DataHandler dataHandler) throws IOException {
        final ByteArrayInputStream is = (ByteArrayInputStream) dataHandler.getContent();
        final int length = is.available();
        final byte[] result = new byte[length];
        is.read(result);
        return result;
    }


    public static DataHandler wrapBytes(final byte[] bytes) {
        return new DataHandler(new DataSource() {

            @Override
            public InputStream getInputStream() throws IOException {
                return new ByteArrayInputStream(bytes);
            }


            @Override
            public OutputStream getOutputStream() throws IOException {
                throw new UnsupportedOperationException("Not supported yet.");
            }


            @Override
            public String getContentType() {
                return "application/octet-stream";
            }


            @Override
            public String getName() {
                throw new UnsupportedOperationException("Not supported yet.");
            }
        });
    }
}

File: zenfolio/src/main/scala/org/podval/photo/zenfolio/Bytes.scala
/*
 *  Copyright 2011 Leonid Dubinsky <dub@podval.org>.
 * 
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 * 
 *       http://www.apache.org/licenses/LICENSE-2.0
 * 
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *  under the License.
 */

package org.podval.photo.zenfolio

import java.security.{MessageDigest, NoSuchAlgorithmException}

import javax.activation.{DataHandler, DataSource}

import java.io.{InputStream, OutputStream, ByteArrayInputStream, IOException}


object Bytes {

    @throws(classOf[NoSuchAlgorithmException])
    def hash(what: Array[Byte]): Array[Byte] = {
        val digest: MessageDigest = MessageDigest.getInstance("SHA-256")
        digest.reset()
        digest.digest(what)
    }


    @throws(classOf[IOException])
    def readBytes(dataHandler: DataHandler): Array[Byte] = {
        val is: ByteArrayInputStream = dataHandler.getContent().asInstanceOf[ByteArrayInputStream]
        val length: Int = is.available()
        val result: Array[Byte] = new Array[Byte](length)
        is.read(result)
        result
    }


    def wrapBytes(bytes: Array[Byte]): DataHandler =  new DataHandler(new BytesDataSource(bytes))
}



private final class BytesDataSource(bytes: Array[Byte]) extends DataSource {

    @throws(classOf[IOException])
    override def getInputStream(): InputStream = new ByteArrayInputStream(bytes)
            

    @throws(classOf[IOException])
    override def getOutputStream(): OutputStream = throw new UnsupportedOperationException("Not supported yet.")


    override def getContentType(): String = "application/octet-stream"


    override def getName(): String = throw new UnsupportedOperationException("Not supported yet.")
}
